### PR TITLE
bis api

### PR DESCRIPTION
**Pull request by remstone7**

### CODE REVIEW

File: Model/Reclaim.php [hunk 1]
```diff
@@ -106,6 +111,52 @@ public function product($quote_id, $item_id) {
...
+    */
+    public function backinstock($product_id)
+    {
+        if (!$product_id){
+            throw new NotFoundException(_('A product id is required'));
+        }
+        $product = $this->_objectManager->create('Magento\Catalog\Model\Product')->load($product_id);
+
+        if (!$product){
+            throw new NotFoundException(_('A product was not found'));
```
Comment (klaviyojad) [inline Model/Reclaim.php]: maybe call out the product_id passed in the Exception

File: Model/Reclaim.php [hunk 2]
```diff
@@ -106,6 +111,52 @@ public function product($quote_id, $item_id) {
...
+            'price' => $product->getPrice(),
+            'available' => true,
+            'inventory_quantity' => $this->_stockItem->getStockQty($productId),
+            'inventory_policy' => $this->_getStockItem($productId)
+        ));
+        // check to see if the product has variants, if it doesn't just return the product information
+        try {
+            $_children = $product->getTypeInstance()->getUsedProducts($product);
+            // throws a fatal error, so catch it generically and return
+        } catch (\Error $e) {
```
Comment (klaviyojad) [inline Model/Reclaim.php]: is that a typo?

File: Model/Reclaim.php [hunk 3]
```diff
@@ -106,6 +111,52 @@ public function product($quote_id, $item_id) {
...
+        $product = $this->_objectManager->create('Magento\Catalog\Model\Product')->load($product_id);
+
+        if (!$product){
+            throw new NotFoundException(_('A product was not found'));
+        }
+
+        $productId = $product->getId();
+
+        $response = array(array(
+            'id' => $productId,
```
Comment (dmitrymrn) [inline Model/Reclaim.php]: image


It looks like Magento's API doesn't support IDs and only supports SKUs. 
https://devdocs.magento.com/guides/v2.3/rest/list.html

Can you please change this to:
```suggestion
            'id' => $product->getSku(),
```

File: Model/Reclaim.php [hunk 4]
```diff
@@ -106,6 +111,52 @@ public function product($quote_id, $item_id) {
...
+        try {
+            $_children = $product->getTypeInstance()->getUsedProducts($product);
+            // throws a fatal error, so catch it generically and return
+        } catch (\Error $e) {
+            return $response;
+        }
+        
+        foreach ($_children as $child){
+            $response['variants'][] = array(
+                'id' => $child->getId(),
```
Comment (dmitrymrn) [inline Model/Reclaim.php]: And here too please:
```suggestion
                'id' => $child->getSku(),
```

File: Model/Reclaim.php (hunk 3 above)
Comment (dmitrymrn) [reply Model/Reclaim.php]: Aargh... 🤕We need a product ID, which we can use to find its SKU. Let's keep this as is. But we still need to pass variant SKU.

Comment (dmitrymrn) [reply Model/Reclaim.php]: image

I think we also need to pass the store ID for both product and variant... 🤢

File: Model/Reclaim.php (hunk 2 above)
Comment (dmitrymrn) [reply Model/Reclaim.php]: This is a way to set the local scope. Without it, PHP would try to find `\Model\Reclaim\Error`

File: Model/Reclaim.php (hunk 3 above)
Comment (remstone7) [reply Model/Reclaim.php]: good call on store id, will make it optional arg

File: Model/Reclaim.php [hunk 5]
```diff
@@ -106,6 +113,59 @@ public function product($quote_id, $item_id) {
...
+            $product = $this->_productFactory->create()->setStoreId($store_id)->load($product_id);    
+        } else {
+            $product = $this->_productFactory->create()->load($product_id);
+        }
+
+        if (!$product){
+            throw new NotFoundException(_('A product with id '. $product_id .' was not found'));
+        }
+
+        $productId = $product->getId();
```
Comment (Jaeto) [inline Model/Reclaim.php]: If `$product_id` is provided in the signature why do you need to get it again? Is this value different?

File: Model/Reclaim.php [hunk 6]
```diff
@@ -235,4 +295,9 @@ public function _getImages($product){
         }
         return $image_array;
     }
+    public function _getStockItem($productId)
+    {
+        $stock = $this->_stockItemRepository->get($productId);
+        return $stock->getManageStock();
```
Comment (Jaeto) [inline Model/Reclaim.php]: do you need the var to return vs just retuning the `this->_stockItemRepository->($productId);`

**Pull request merged**